cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1155 @@
+		}
+	}
+	s.currentTs = currentTs
+	// TODO: is it necessary to print changes detailly?
+	log.Debug("adjust partition success",
+		zap.String("schema", tbInfo.TableName.Schema),
+		zap.String("table", tbInfo.TableName.Table))
+	return nil
+}
+
+func (s *Snapshot) renameTables(job *timodel.Job, currentTs uint64) error {
+	var oldSchemaIDs, newSchemaIDs, oldTableIDs []int64
+	var newTableNames, oldSchemaNames []*timodel.CIStr
+	err := job.DecodeArgs(&oldSchemaIDs, &newSchemaIDs, &newTableNames, &oldTableIDs, &oldSchemaNames)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(job.BinlogInfo.MultipleTableInfos) < len(newTableNames) {
+		return cerror.ErrInvalidDDLJob.GenWithStackByArgs(job.ID)
+	}
+	// NOTE: should handle failures in halfway better.
+	for _, tableID := range oldTableIDs {
+		if err := s.dropTable(tableID, currentTs); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	for i, tableInfo := range job.BinlogInfo.MultipleTableInfos {
+		newSchema, ok := s.SchemaByID(newSchemaIDs[i])
+		if !ok {
+			return cerror.ErrSnapshotSchemaNotFound.GenWithStackByArgs(newSchemaIDs[i])
+		}
+		newSchemaName := newSchema.Name.L
+		tbInfo := model.WrapTableInfo(newSchemaIDs[i], newSchemaName, job.BinlogInfo.FinishedTS, tableInfo)
+		err = s.createTable(tbInfo, currentTs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// HandleDDL handles the given job.
+func (s *Snapshot) HandleDDL(job *timodel.Job) error {
+	if err := s.FillSchemaName(job); err != nil {
+		return errors.Trace(err)
+	}
+	return s.DoHandleDDL(job)
+}
+
+// Drop drops the snapshot. It must be called when GC some snapshots.
+// Drop a snapshot will also drop all snapshots with a less timestamp.
+func (s *Snapshot) Drop() {
+	if s.lock(false) {
+		defer s.unlock()
+	}
+	tag := negative(s.currentTs)
+
+	schemas := make([]versionedID, 0, s.schemas.Len())
+	var schemaID int64 = -1
+	var schemaDroped bool = false
+	s.schemas.Ascend(func(i btree.Item) bool {
+		x := i.(versionedID)
+		if x.tag >= tag {
+			if x.id != schemaID {
+				schemaID = x.id
+				schemaDroped = false
+			}
+			if schemaDroped || x.target == nil {
+				schemas = append(schemas, newVersionedID(x.id, x.tag))
+			}
+			schemaDroped = true
+		}
+		return true
+	})
+	for _, vid := range schemas {
+		s.schemas.Delete(vid)
+	}
+
+	tables := make([]versionedID, 0, s.tables.Len())
+	var tableID int64 = -1
+	var tableDroped bool = false
+	s.tables.Ascend(func(i btree.Item) bool {
+		x := i.(versionedID)
+		if x.tag >= tag {
+			if x.id != tableID {
+				tableID = x.id
+				tableDroped = false
+			}
+			if tableDroped || x.target == nil {
+				tables = append(tables, newVersionedID(x.id, x.tag))
+			}
+			tableDroped = true
+		}
+		return true
+	})
+	for _, vid := range tables {
+		x := s.tables.Delete(vid).(versionedID)
+		info := (*model.TableInfo)(x.target)
+		if info != nil {
+			isEligible := info.IsEligible(s.forceReplicate)
```

Comment:
`ineligible := !info.IsEligible(s.forceReplicate)` may be more easy to understand.